Custom Restaurant Kitchen Management System Django Project with basic front-end

tylerj231 opened the pull request.

TylerJack239 committed:
Add global project settings. Create application

Added global project settings. Created application with the following models: Dish, DishType, Cook.

TylerJack239 committed:
Add ability to view list of dishes/dish_types/cooks & detail view for specific dish

Created respective url paths, views & templates for dish, dish_type & cooks. Now there is an ability to view all available dishes and get more details about specific dish. For cooks only list view is available at the moment.

TylerJack239 committed:
Add basic built-in Django authentication. Add ability to create dish/dish_type & cook

Added basic built-in Django authentication with the ability to log in & log out.  Implemented the ability to create a dish/dish_type & cook through adding
appropriate urls, views & templates.

TylerJack239 committed:
Add ability to update & delete dish & cook.

Added the ability to update and delete specific dish and cook through impelemnting respective views, urls & templates where needed. Before deletion of
either a dish or a cook user has to confirm his choice, then the item will be deleted.

TylerJack239 committed:
Modified project settings & front-end side of application

Minor changes to project settings regarding static files. Improved front-end side of the application across all app templates by using
bootstrap5 framework & crispy forms

TylerJack239 committed:
Add minor changes to models, views, admin. Add test for models, views, admin

Added tests for cook, dish & dish_type views that check if list & detail page are correctly retreived for the logged-in user
Added test for admin panel that check if the "years of experience" field is correctly displayed in various cases.
Added test for models to check if the redefined str representation is returned as expected & whether a cook with "years of experience" field is created correctly.
Removed "LoginRequiredMixin" from HomePageView.
Removed verbose & verbose_pluarl attributes DishType

TylerJack239 committed:
Change verbose_name_plural for cook & dish to the correct one. Format style of views.py

Change verbose_name_plural attribute from cook -> cooks & dish -> dishes. Added one line break in views.py

TylerJack239 committed:
Remove unnecessary code from views.py. Fix issue with UpdateView for DishType

Removed .edit in the parenthesis of the class from every Update & Delete view. Fixed issue with the Update view for DishType by changing inheritence from Detail view to UpdateView class

tylerj231 commented:
![Home Page](https://github.com/user-attachments/assets/d7c3bce6-9559-4667-9a97-ee4cc02d32ab)
![dish delete](https://github.com/user-attachments/assets/7b3af488-9539-440a-9e65-2a16e45f525a)
![dish update](https://github.com/user-attachments/assets/ac02a347-dd54-4bca-bd21-4e6853382280)
![dish detail](https://github.com/user-attachments/assets/ec276e9a-d909-4501-a66b-43594efecc05)
![Dish list](https://github.com/user-attachments/assets/c5115f14-fa99-426d-83e1-c27bba44e4cf)
![dish create](https://github.com/user-attachments/assets/9609385e-09b7-4af0-8606-1fef5fbf5d26)
![dish type delete](https://github.com/user-attachments/assets/68dc5c22-554e-4b01-8c02-bf5a6fe4c9fd)
![dish type update](https://github.com/user-attachments/assets/a5731507-2334-4ca8-bf68-e8bb8000d77a)
![dish create](https://github.com/user-attachments/assets/e68b8262-e2ba-49b6-aedd-9b08ec3f32f1)
![dish type list](https://github.com/user-attachments/assets/864c77d4-d220-4840-abe3-cc6cf458a02a)
![cook create](https://github.com/user-attachments/assets/06737ddd-3780-411f-b025-bab8a049d018)
![cook delete](https://github.com/user-attachments/assets/468ebfe0-cb66-408f-b8d1-c6650cdc7fe0)
![cook update](https://github.com/user-attachments/assets/fd5c3b89-6e06-49e9-8572-8276ff81228d)
![cook detail](https://github.com/user-attachments/assets/fbd15c8e-defd-42b8-b636-527ac14d2ba3)
![cook list](https://github.com/user-attachments/assets/b1e975cf-cbb3-4be9-9baa-bba12cd95aee)
![log in page](https://github.com/user-attachments/assets/7dbd486a-8a19-4e31-907e-ee5f40afd8af)
![DB structure](https://github.com/user-attachments/assets/dce3ec7a-af54-460d-b604-e25365813eb9)

OlegAndrus12 commented:
Some remarks from user perspective:
1. **cooks details page**: 
  - add max-width for the image, otherwise it could occupy the whole screen. 
  - Also if user did not specify how many years of experience he has then just don't display it, no need for showing None.
2. **update cook page**: 
  - never show the password on a frontend, even after authentication. Just reset password form. 
  - Also, there is no use of generated hash for the user.
 3. **add favicon**
 4. **cooks/<id>/delete**: 
   - confirmation box is too wide
 5. **cooks create page**:
   - you have to mention password requirements, same way to do for the username
  6. **dish_type page**:
    - update/delete buttons should be always right aligned. currently update is under details column, since details are not provided. Make them sticky to the right
  7. **homepage might need some more work, looks poorly, at least text-center the numbers**

In general I'm glad you figured out how to use bootstrap components.

OlegAndrus12 commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+.env
+.idea
+db.sqlite3
+**/__pycache__/
```
empty line in the end of the file

OlegAndrus12 commented on KitchenService/__pycache__/__init__.cpython-312.pyc:
aren't they ignored?

tylerj231 replied:
Fixed, deleted those files from git repo

OlegAndrus12 commented on images/Crazy-Dave-Costume_600x.jpg:
shouldn't be hardcoded

tylerj231 replied:
fixed. Configured handling of the static files properly. Now, I use the image through {%static url ''%}

OlegAndrus12 commented on kitchen_service/settings.py:
```diff
@@ -0,0 +1,102 @@
+
+load_dotenv()
+
+BASE_DIR = Path(__file__).resolve().parent.parent
```
it might work without .resolve

tylerj231 replied:
I tried, it does not work that way, in my case at least.

OlegAndrus12 commented on management_system/admin.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@admin.register(Dish)
+class DishModelAdmin(admin.ModelAdmin):
```
search fields?

OlegAndrus12 replied:
it would be nice to have a search bar in django admin

tylerj231 replied:
Added a search bar for dishes, dish types & cooks.

OlegAndrus12 commented on management_system/models.py:
```diff
@@ -0,0 +1,36 @@
+    dish_type = models.ForeignKey(
+        "DishType", on_delete=models.CASCADE, related_name="dishes"
+    )
+    cooks = models.ManyToManyField(settings.AUTH_USER_MODEL, related_name="dishes")
```
i'd make it vice versa, for example Cook has a list of the dishes he is qualified to cook, not the Dish has the list of cooks. Most of the time you will have to get dishes based on the cook, not the other way

tylerj231 replied:
Thank you, I will change the relationship as you suggested a bit later.

OlegAndrus12 commented on management_system/models.py:
```diff
@@ -0,0 +1,36 @@
+
+
+class DishType(models.Model):
+    name = models.CharField(max_length=255)
```
can you imagine dish type 255 characters long?

tylerj231 replied:
Shortened to 55 char max.

OlegAndrus12 commented on management_system/tests/test_admin/test_admin.py:
```diff
@@ -0,0 +1,42 @@
+        :return:
+        """
+        url = reverse("admin:management_system_cook_changelist")
+        response = self.client.get(url)
```
check for the status code too

OlegAndrus12 replied:
in all the tests

OlegAndrus12 replied:
what if returns years=2 when the cook has 25 (according to the setup). assertContains is not enough, also check the value

tylerj231 replied:
fixed.

OlegAndrus12 commented on management_system/tests/test_models/test_models.py:
```diff
@@ -0,0 +1,62 @@
+        password = "test12333"
+        years_of_experience = 10
+
+        cook = get_user_model().objects.create_user(
```
move get_user_model outside of the class
`User = get_user_model()`

tylerj231 replied:
Done.

tylerj231 replied:
Done.

OlegAndrus12 commented on management_system/tests/test_views/test_cook_view.py:
```diff
@@ -0,0 +1,46 @@
+        Test that checks if the cook list view is returned for the logged-in user
+        :return:
+        """
+        cook_1 = Cook.objects.create(
```
add more cooks and use `bulk_create` method

docs https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create

tylerj231 replied:
Done.

OlegAndrus12 commented on management_system/tests/test_views/test_dish_view.py:
```diff
@@ -0,0 +1,43 @@
+
+
+class PrivateDishPageTestCase(TestCase):
+    def setUp(self):
```
if you were using function based tests, you could create authenticated_user fixture once and use it for all the tests instead of writing the same presetup

tylerj231 replied:
Thank you, I will try this approach of writing tests for my next project.

OlegAndrus12 commented on templates/cook/cook_confirm_delete.html:
```diff
@@ -0,0 +1,22 @@
+    </div>
+  </div>
+
+{% endblock %}
```
end of file

OlegAndrus12 commented on templates/cook/cook_confirm_delete.html:
```diff
@@ -0,0 +1,22 @@
+          {% csrf_token %}
+          {{ form|crispy }}
+          <div class="d-flex justify-content-end">
+            <a href="{% url 'management_system:cooks-list' %}" class="btn btn-secondary me-2">Cancel</a>
```
me-2 class does not exist, i think you made a typo, probably mt-2 is correct

tylerj231 replied:
fixed.

OlegAndrus12 commented on templates/cook/cook_detail.html:
```diff
@@ -0,0 +1,19 @@
+
+  <div class="container my-5">
+    <div class="card">
+      <img src="https://qudahalloween.com/cdn/shop/articles/Crazy-Dave-Costume_600x.png?v=1719396046"
```
add width and height properties, or at least max-width

tylerj231 replied:
fixed.

OlegAndrus12 commented on templates/cook/cook_form.html:
```diff
@@ -0,0 +1,16 @@
+    </form>
+  </div>
+
+{% endblock %}
```
end of file, check other files too pls

tylerj231 replied:
Fixed in all other files.

OlegAndrus12 commented on templates/cook/cook_list.html:
```diff
@@ -0,0 +1,40 @@
+      <a href="{% url 'management_system:cook-create' %}" class="btn btn-primary">Create Cook</a>
+    </div>
+
+    {% if cook_list %}
```
if cook_list is not provided just display an empty table

OlegAndrus12 replied:
no need for if

tylerj231 replied:
Removed if else statement.

OlegAndrus12 commented on templates/cook/cook_list.html:
```diff
@@ -0,0 +1,40 @@
+          </thead>
+          <tbody>
+          {% for cook in cook_list %}
+            <tr>
```
also add a column for ordinal number, so user see how many records in the table (starting by 1)

tylerj231 replied:
Thanks, I will implement this feature, a little bit later on.

TylerJack239 committed:
Remove __pycache__ files from Git tracking

TylerJack239 committed:
Add Search functionality. Refactor code based on OlegAndrus12 feedback.

Added search form to cooks, dish & dish type lists, so that user can search a particular item based on certain field. Added custom search forms, respectively.
Removed py_cache files from git repo, put them into .gitignore file.
Refactored, models, tests, views.

Properly configured  handling of static files.
Squashed multiple migrations into one according to mentor’s feedback as this is a best practise.

TylerJack239 committed:
Complete squishing migrations.

tylerj231 committed:
Delete __pycache__ directory
